### DISCUSSION TITLE
Metric Units & Descriptions

### DISCUSSION THREAD

**Discussion by paulevanmr** [open-telemetry/opentelemetry-dotnet#5986] [Q&A]
When registering my instrument I am adding a unit and a description and my expectation would be that somewhere within the metric payload it contains these values however that is not the case.

For context: my setup is: 

.NET OTEL SDK > OTEL Collector > Elastic APM > Elasticsearch > Visualised in Kibana

Below is an example payload coming out of one of my applications (I've redacted a few values so ignore the X's). 

```
{
  "_index": ".ds-metrics-business-environment-2024.11.18-000001",
  "_id": "aVCpP5MBnbd0pV9RaW1x",
  "_version": 1,
  "_score": 0,
  "_source": {
    "observer": {
      "hostname": "fef008961d3c",
      "type": "apm-server",
      "version": "8.15.3"
    },
    "agent": {
      "name": "opentelemetry/dotnet",
      "version": "1.9.0"
    },
    "@timestamp": "2024-11-18T14:25:54.809Z",
    "data_stream": {
      "namespace": "X",
      "type": "metrics",
      "dataset": "business"
    },
    "service": {
      "node": {
        "name": "40189601-80eb-49db-9e71-9c28e6ec577e"
      },
      "environment": "X",
      "framework": {
        "name": "X.meter"
      },
      "name": "X",
      "language": {
        "name": "dotnet"
      },
      "version": "2.1.0"
    },
    "host": {
      "id": "X"
    },
    "metricset": {
      "name": "app",
      "samples": [
        {
          "name": "X",
          "type": "counter",
          "value": 3
        }
      ]
    },
    "event": {},
    "labels": {
      "deployment_environment_name": "x",
      "X": "X",
      "metric_kind": "business"
    }
  },
  "fields": {
    "service.environment": [
      "X"
    ],
    "service.name": [
      "X"
    ],
    "service.framework.name": [
      "X.meter"
    ],
    "data_stream.namespace": [
      "X"
    ],
    "labels.deployment_environment_name": [
      "X"
    ],
    "labels.X": [
      "X"
    ],
    "service.node.name": [
      "40189601-80eb-49db-9e71-9c28e6ec577e"
    ],
    "data_stream.type": [
      "metrics"
    ],
    "observer.hostname": [
      "fef008961d3c"
    ],
    "service.language.name": [
      "dotnet"
    ],
    "metricset.name": [
      "app"
    ],
    "metricset.samples.name": [
      "X"
    ],
    "@timestamp": [
      "2024-11-18T14:25:54.809Z"
    ],
    "service.version": [
      "2.1.0"
    ],
    "observer.type": [
      "apm-server"
    ],
    "observer.version": [
      "8.15.3"
    ],
    "data_stream.dataset": [
      "business"
    ],
    "agent.name": [
      "opentelemetry/dotnet"
    ],
    "agent.version": [
      "1.9.0"
    ],
    "metricset.samples.type": [
      "counter"
    ],
    "metricset.samples.value": [
      3
    ],
    "host.id": [
      "X"
    ],
    "labels.Xmetric_kind": [
      "business"
    ]
  }
}
```

No where in here does it mention a unit or description. Looking at the payload I would expect it maybe somewhere here:

```
"metricset": {
      "name": "app",
      "samples": [
        {
          "name": "X",
          "type": "counter",
          "value": 3
        }
      ]
    },
```

I haven't found anything in the documentation that specifically states the unit / description should or shouldn't be part of the payload, but I don't see why it would be there if it wasn't intended to be?

**Comment by cijothomas** [comment]
> .NET OTEL SDK > OTEL Collector > Elastic APM > Elasticsearch > Visualised in Kibana

For upto **_NET OTEL SDK > OTEL Collector_**, unit, description are part of. Beyond that, you can consult Elastic/Kibana experts.

**Reply by paulevanmr** [reply]
> > .NET OTEL SDK > OTEL Collector > Elastic APM > Elasticsearch > Visualised in Kibana
> 
> For upto **_NET OTEL SDK > OTEL Collector_**, unit, description are part of. Beyond that, you can consult Elastic/Kibana experts.

Thanks for the reply @cijothomas - I am not expecting the elastic aspect to be answered; was merely giving some additional context. 

My question remains the same regardless: 

Should I expect unit & description to be a part of the payload that comes out of the dotnet SDK.

Thanks

**Reply by cijothomas** [reply]
_Should I expect unit & description to be a part of the payload that comes out of the dotnet SDK._

Yes. To be precise, its part of the data that SDK hands over to the MetricExporters, and it is entirely upto the exporter to decide what to do with these information. The Console and OTLP Exporter from this repo already handles them. I am about ~99% sure that Prometheus exporter also takes care of them.
